smtp-listeners/src/main/kotlin/no/nav/emottak/smtp/cpasync/CpaSyncService.kt

```diff
@@ -0,0 +1,103 @@
+package no.nav.emottak.smtp.cpasync
+
+import com.jcraft.jsch.SftpException
+import io.ktor.client.HttpClient
+import no.nav.emottak.deleteCPAinCPARepo
+import no.nav.emottak.getCPATimestamps
+import no.nav.emottak.nfs.NFSConnector
+import no.nav.emottak.putCPAinCPARepo
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.time.Instant
+import java.util.*
+
+class CpaSyncService(private val cpaRepoClient: HttpClient) {
+
+    private val log: Logger = LoggerFactory.getLogger("no.nav.emottak.smtp.cpasync")
+
+    suspend fun sync(nfsConnector: NFSConnector = NFSConnector()) {
+        return runCatching {
+            val mutableCpaTimestamps = cpaRepoClient.getCPATimestamps().toMutableMap()
+            processAndSyncEntries(mutableCpaTimestamps, nfsConnector)
+            deleteStaleCpaEntries(mutableCpaTimestamps)
+        }.onFailure {
+            logFailure(it)
+        }.getOrThrow()
+    }
+
+    private suspend fun processAndSyncEntries(
+        mutableCpaTimestamps: MutableMap<String, String>,
+        nfsConnector: NFSConnector
+    ) {
+        nfsConnector.use { connector ->
+            connector.folder().forEach { entry ->
+                val filename = entry.filename
+                log.info("Checking $filename...")
+                if (!filename.endsWith(".xml")) {
+                    log.warn("$filename is ignored")
+                    return@forEach
+                }
+
+                val lastModified = Date(entry.attrs.mTime.toLong() * 1000).toInstant()
+                if (shouldSkipFile(filename, lastModified, mutableCpaTimestamps)) {
+                    return@forEach
+                }
+
+                runCatching {
+                    log.info("Fetching file ${entry.filename}")
+                    val cpaFileContent = connector.file("/outbound/cpa/${entry.filename}").use {
+                        String(it.readAllBytes())
+                    }
+                    log.info("Uploading $filename")
+                    cpaRepoClient.putCPAinCPARepo(cpaFileContent, lastModified)
+                }.onFailure {
+                    log.error("Error uploading $filename to cpa-repo: ${it.message}", it)
+                }
+            }
+        }
+    }
+
+    internal fun shouldSkipFile(
+        filename: String,
+        lastModified: Instant,
+        mutableCpaTimestamps: MutableMap<String, String>
+    ): Boolean {
+        val shouldNotBeDeleted = ArrayList<String>()
+        val matches = mutableCpaTimestamps.filter { (cpaId, timestamp) ->
+            val formattedCpaId = cpaId.replace(":", ".")
+            if (filename.contains(formattedCpaId)) {
+                shouldNotBeDeleted.add(cpaId)
+                if (Instant.parse(timestamp) == lastModified) {
```

Comment:
OBS, alle systemer har ikke samme presisjon på timestamp. (.0001 vs .00011 type problemstilling, så det må tas høyde for, jeg var nok litt vel aggressiv med 2 sekunders slingringsmonnet som var, men forskjell på millisekunder vil gi timestamp mismatch.

Reply:
Forstår jeg dette riktig:
Sier du at hvis vi lagrer noe i vår database, og så prosesserer vi nøyaktig samme fil igjen, så blir vi alltid å få false her?

Reply:
Ja det vil skje (har skjedd).

Derfor har jeg lagt til litt ekstra funksjonalitet som truncater millisekundpresisjon på lagring av timestamp i CPA repo sånn at man kan kanskje sjekke på sekund isteden. Jeg var litt ekstra slakk og sa at om den var oppdatert innen 2 sekunder siden sist så er den oppdatert. (Spoiler: jeg mistenker at dette er noe av årsak til "buggen".)

Reply:
Lager tester for dette i morgen, hvis vi har lagret ms, kan jeg begrense det til de 2 eller 3 siste ms. Så lenge vi lagrer 4 (eller større) presisjon burde det bli rett så lenge jeg får verifisert logikken via testene.